fmriprep/reports/core.py

```diff
@@ -54,10 +75,52 @@ def generate_reports(subject_list, output_dir, run_uuid, config=None, work_dir=N
             import sys
             import traceback
 
+            # Store the list of subjects for which report generation failed
             errors.append(subject_label)
             traceback.print_exception(
                 *sys.exc_info(),
                 file=str(Path(output_dir) / "logs" / f"report-{run_uuid}-{subject_label}.err"),
             )
 
+        if n_ses >= config.execution.max_ses_agr:
+            # Beyond a certain number of sessions per subject, we separate the functional reports per session
+            if session_list is None:
+                session_list = config.execution.layout.get_sessions(subject=subject_label)
+
+            # Drop ses- prefixes
+            session_list = [ses[4:] if ses.startswith("ses-") else ses for ses in session_list]
+
+            for session_label in session_list:
+                bootstrap_file = data.load("reports-spec-func.yml")
+                html_report = ''.join(
+                    [f"sub-{subject_label}", f"_ses-{session_label}", "_func.html"]
+                )
+                entities["session"] = session_label
+
+                robj = Report(
+                    output_dir,
+                    run_uuid,
+                    bootstrap_file=bootstrap_file,
+                    out_filename=html_report,
+                    reportlets_dir=reportlets_dir,
+                    plugins=None,
+                    plugin_meta=None,
+                    metadata=None,
+                    **entities,
+                )
+
+                try:
+                    robj.generate_report()
+                except:
+                    # Store the list of subjects for which report generation failed
+                    errors.append(subject_label)
+                    traceback.print_exception(
+                        *sys.exc_info(),
+                        file=str(
+                            Path(output_dir)
+                            / "logs"
+                            / f"report-{run_uuid}-{subject_label}-func.err"
+                        ),
+                    )
```

Comment:
This should probably be pulled out into a little function instead of repeated from above. Otherwise we risk getting out of sync.